libsqlite3-sys/src/lib.rs

```diff
@@ -56,6 +56,25 @@ pub enum Limit {
     SQLITE_LIMIT_WORKER_THREADS = 11,
 }
 
+#[cfg(feature = "loadable_extension_embedded")]
+// bindings were built with loadable_extension_embedded:
+// define sqlite3_api as an extern since this code will be embedded
+// within a loadable extension that defines and exports this itself
+extern "C" {
+    pub static mut sqlite3_api: *mut sqlite3_api_routines;
+}
```

Comment:
@thomcc commented on this in the now closed PR: https://github.com/rusqlite/rusqlite/pull/868#discussion_r546279767
"This should probably be an AtomicPtr, and have a Release store and Acquire loads. Otherwise it's a data race."

Reply:
This was discussed against the earlier PR: https://github.com/rusqlite/rusqlite/pull/631#discussion_r416921802

Reply:
I will attempt the proposed addition of `sqlite_extension_init` that I suggested in https://github.com/rusqlite/rusqlite/pull/631#discussion_r416921802

Reply:
This has now been implemented as `loadable_extension_init` and `loadable_extension_embedded_init`. 

@thomcc are you happy with this?

Reply:
It seems like you use a once for the write and not for the reads. This is not enough for happens-before. You need an acquire on the reader threads, or for them to attempt to run the `Once` too (which they can't, because they don't the ptr to initialize. So it has a data race.

Instead, storing it with an `AtomicPtr` with a `Release` store, and reading it with an `Acquire` load would be better.

But I think this should be done fairly differently, as mentioned.

Reply:
@thomcc Something likes [that](https://gist.github.com/gwenn/32beb99dec960f73e025e8b322199812) ?

Reply:
You need at least a Release store (and Acquire loads) for proper happens-before.

Reply:
Thanks !
And another question: how to handle [va_list](https://github.com/rusqlite/rusqlite/blob/master/libsqlite3-sys/sqlite3/sqlite3ext.h#L160) ?
I guess we cannot remove this field (`vmprintf`) from the generated bindings (like we currently do for the associated function `sqlite3_vmprintf`).
I've tried to use bindgen [opaque_type("(?i).*va_list.*")](https://docs.rs/bindgen/0.66.1/bindgen/struct.Builder.html#method.opaque_type) without success.
So temporary, I've manually replaced `va_list` by `void*`.

Reply:
@thomcc I manage to generate a `bindgen_bundled_version_ext.rs` with [this](https://gist.github.com/gwenn/dbd03af8c4189cf7acc765348aa17893#file-main-rs) code.
```rust
...
static __SQLITE3_LAST_INSERT_ROWID: ::std::sync::atomic::AtomicPtr<
    unsafe extern "C" fn(arg1: *mut sqlite3) -> sqlite_int64,
> = ::std::sync::atomic::AtomicPtr::new(std::ptr::null_mut());
pub unsafe fn sqlite3_last_insert_rowid(arg1: *mut sqlite3) -> sqlite_int64 {
    let fun = __SQLITE3_LAST_INSERT_ROWID.load(::std::sync::atomic::Ordering::Acquire);
    assert!(! fun.is_null(), "SQLite API not initialized");
    (*fun)(arg1)
}
...

/// Like SQLITE_EXTENSION_INIT2 macro
pub unsafe fn rusqlite_extension_init2(p_api: *mut sqlite3_api_routines) {
    assert!(! p_api.is_null(), "Invalid sqlite3_api_routines pointer");
...
    __SQLITE3_LAST_INSERT_ROWID
        .store(
            &mut (*p_api).last_insert_rowid.unwrap(),
            ::std::sync::atomic::Ordering::Release,
        );
...
}
```
Could you please tell me if this is the right way ?
Thanks.

Reply:
I think instead of `&mut (*p_api).last_insert_rowid.unwrap()` you should store the function address directly, since the result of `.last_insert_rowid.unwrap()` is a temporary, and you're storing a pointer to the result of a temporary. I'm a little surprised it works.

There are a few ways you could fix it but the way with the best performance would be https://gist.github.com/thomcc/242dcdcc3fa840567dd4ea7c75ee896c probably. Let me know if that makes sense to you.

I don't think there's anything we can do for functions with `va_list` or `...`

Reply:
@thomcc 
(a) What about the [atomic](https://crates.io/crates/atomic) crate which gives [this](https://gist.github.com/gwenn/32beb99dec960f73e025e8b322199812) ?
```rust
static _sqlite3_db_config: Atomic<
    Option<unsafe extern "C" fn(arg2: ::std::os::raw::c_int, ...) -> ::std::os::raw::c_int>,
> = Atomic::new(None);
```
No need to transmute and is lock free:
```rust
    println!(
        "is_lock_free: {}",
        Atomic::<
            Option<unsafe extern "C" fn(arg2: ::std::os::raw::c_int, ...) -> ::std::os::raw::c_int>,
        >::is_lock_free()
    ); // is_lock_free: true
```

(b) And for `va_list`, how can we make the generated bindings "portable" for:
```c
struct sqlite3_api_routines {
...
  char *(*vmprintf)(const char*,va_list);
...
  char *(*xvsnprintf)(int,char*,const char*,va_list);
...
  void (*str_vappendf)(sqlite3_str*, const char *zFormat, va_list);
...
};
```
?
Thanks.

Reply:
The atomic crate looks fine if used as you describe. I think the other stuff is not possible to define correctly in Rust at the moment (prior to `core::ffi::VaList` stabilizing), but you could just the function pointers filled out with `*const ()` or something, as a placeholder. There's no real reason for people to use those functions in Rust anyway.

Reply:
Ok,
I am able to do a:
```sh
rusqlite % cargo check --features "modern_sqlite,loadable_extension,vtab,trace"
```
by manually fixing the signature of `sqlite3_db_config` and `sqlite3_log` (see https://gist.github.com/gwenn/dbd03af8c4189cf7acc765348aa17893#file-main-rs-L135-L152).
Then replacing locally the default bindings by the one generated by https://gist.github.com/gwenn/dbd03af8c4189cf7acc765348aa17893#file-main-rs :
```sh
rusqlite % cp ../sqlite-ext/bindgen_bundled_version_ext.rs libsqlite3-sys/sqlite3/bindgen_bundled_version.rs
```
And applying this minimal patch: https://gist.github.com/gwenn/dbd03af8c4189cf7acc765348aa17893#file-rusqlite-patch.
So the last remaining step is fixing libsqlite3-sys/build.rs by integrating https://gist.github.com/gwenn/dbd03af8c4189cf7acc765348aa17893#file-main-rs...